### DISCUSSION TITLE
how to limit SharedMemory feature to localhost only

### DISCUSSION THREAD

**Discussion by keith4ever** [eProsima/Fast-DDS#2721] [Q&A] [answered]
Hello,

I'm using FastDDS on an ARM64 based embedded system. 
I need to use sharedmem as a requirement and it works well so far. 
But as I scale up, I found out that the sharedmem data is even distributed to other targets in the same LAN, and this clogs the whole network. 

So I wonder if there is a way to limit sharedmem to localhost only..

**Comment by EduPonz** [comment, marked as answer]
Hi @keith4ever ,

I'm not sure completely sure I understand what you mean, but I'll make a guess. I take you want to limit DDS communication to the host, and that you do not want data to be exchanged between different machines because you have somehow duplicated nodes that are not meant to communicate with one another, is that correct?

Being the shared memory transport (SHM) an intra-machine IPC mechanism, it usage is always limited to the local host. However, Fast DDS Domain Participants have, by default, two different transports enabled: SHM and UDPv4. This is the reason why you're seeing data going into your LAN. To achieve machine communication isolation, you have several options:

* Have dedicated [DDS domains](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/domain/domain.html) for the different machines. Mind that this approach will not prevent some discovery going over the LAN using multicast, but it has advantage of being able to tap into one machine's topics from a different host at will. With this method, you achieve the isolation via the discovery protocol, as Domain Participants running in different domains will not discover each other.
* Use [Ignore Participant Flags](https://fast-dds.docs.eprosima.com/en/latest/fastdds/discovery/general_disc_settings.html#discovery-ignore-flags) to avoid participant matching even when they are running on the same domain.
* Use [interface whitelisting](https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/whitelist.html) to limit the network interfaces that Fast DDS is allowed to use. By selecting only localhost, you'd avoid traffic over your LAN.
* [Remove the default transports](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/eprosimaExtensions.html#transportconfigqos) and [add only a SHM](https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/shared_memory/shared_memory.html) to your participants. This way, discovery happens over the SHM transport, so no data can ever get out of the machine.

Hope that helps!

**Reply by jespersmith** [reply]
I'm also trying to enforce shared memory only. Using the 4th options, it seems that without a UDP transport, I'm not able to discover the other participants.

My participant XML is:

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<dds xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
    <profiles>
        <transport_descriptors>
            <transport_descriptor>
                <transport_id>ac320273-2422-4737-a44c-d3e8d2bd6faa</transport_id>
                <type>SHM</type>
            </transport_descriptor>
        </transport_descriptors>
        <participant profile_name="TestProfile">
            <domainId>0</domainId>
            <rtps>
                <builtin>
                    <discovery_config>
                        <leaseDuration>
                            <nanosec>-1</nanosec>
                            <sec>2147483647</sec>
                        </leaseDuration>
                    </discovery_config>
                </builtin>
                <userTransports>
                    <transport_id>ac320273-2422-4737-a44c-d3e8d2bd6faa</transport_id>
                </userTransports>
                <useBuiltinTransports>false</useBuiltinTransports>
                <name>nmpc_interface_node</name>
            </rtps>
        </participant>
    </profiles>
</dds>
```

**Comment by keith4ever** [comment]
Hello EduPonz,

Thank you very much for your kind and detailed answer. 
I have tried the last suggestion of yours and indeed it is broadcast within localhost processes. 
I have a couple of follow up questions if you don't mind:
1. SHM has to be either on UDP or TCP (not transport protocol but base protocol), correct? Can you tell me which base protocol?
2. Can you confirm SHM is zero copy, so it doesn't matter how many subscribers is reading it?
